### PR TITLE
Feat/fetch lfs

### PR DESCRIPTION
**Pull request by volvogustaf**
# Git LFS Support ( #278 ) - Partial Implementation (`git toprepo lfs fetch`)

 
## Overview

 
This PR implements partial Git LFS support for git-toprepo emulated monorepos, specifically the `git toprepo lfs fetch` command.
 
### 1. `git toprepo lfs fetch <path>` Command
- **Core functionality**: Fetches LFS objects for specified paths by resolving the correct remote URL based on submodule configuration
- **Path resolution**: 
  - Maps worktree paths to the appropriate repository (top-level or submodule)
  - Uses `.gitmodules` to find the deepest matching submodule for a given path
  - Resolves relative submodule URLs against the top repository's remote URL
  - Validates that submodule URLs are configured in `.gittoprepo.toml`
- **Supported options**: `--dry-run`, `--prune`, `--recent`, `--refetch`, `--exclude`/-X
- **Path handling**:
  - Supports both absolute and relative paths
  - Handles paths with spaces correctly
  - Works from any subdirectory within the worktree
  - Fetches multiple paths sequentially (one `git lfs fetch` call per path)
 
### 2. LFS Filter Warning System
- **Automatic detection**: During `git toprepo fetch`, checks if LFS filters are configured
- **Warning conditions**:
  - Warns if `filter.lfs.smudge` is set to plain `git-lfs` instead of `git-toprepo lfs smudge`
  - Warns if `filter.lfs.process` is set to plain `git-lfs filter-process` instead of going through git-toprepo
- **Rationale**: If LFS filters bypass git-toprepo, they will fetch from the wrong remote in an emulated monorepo context
 
### 3. New Module: `src/lfs.rs` 
Core LFS functionality including:
- `resolve_lfs_fetch_targets()`: Maps paths to repository URLs
- `run_lfs_fetch()`: Executes `git lfs fetch` with the correct remote
- `warn_if_lfs_filters_bypass_toprepo()`: Checks git config for problematic LFS filter settings
- `ensure_git_lfs_available()`: Validates that Git LFS is installed
- `is_lfs_filter_through_toprepo()`: Parses filter commands to detect git-toprepo wrappers
 
### 4. Comprehensive Test Coverage
- **Unit tests** in `src/lfs.rs`:
  - Filter command parsing
  - Deepest submodule path resolution
  - Path normalization edge cases
- **Integration tests** in `tests/integration/lfs.rs` (362 lines):
  - Fetching from top-level repository
  - Fetching from submodules (resolves correct URL)
  - Relative path handling from subdirectories
  - Multiple path fetching
  - Nested submodules (deepest match wins)
  - Spaces in file paths
  - Error handling for missing Git LFS
  - Error handling for unconfigured subrepositories
 
### Unsupported Options
The following `git lfs fetch` options are explicitly rejected:
- `--all`: Would require fetching from all configured subrepos, complex to implement correctly
- `--stdin`: Not implemented
- `--include`/`-I`: Reserved for internal use (git-toprepo supplies this)
- `--json`: Not implemented

 
## Testing
 
All new code is covered by:
- Unit tests for core logic (filter parsing, path resolution)
- Integration tests using a fake `git-lfs` binary to verify correct command invocation
- Tests cover top-level files, submodule files, nested submodules, error cases
 
Tests only run on Unix due to shell script fixtures.

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -193,6 +193,78 @@ pub enum Commands {
...
+
+#[derive(Subcommand, Debug)]
+pub enum Lfs {
+    /// Fetch LFS objects for monorepo paths.
+    Fetch(LfsFetch),
+}
+
+#[derive(Args, Debug)]
+pub struct LfsFetch {
+    /// Print what Git LFS would fetch, without downloading objects.
```
Comment (stagnation) [inline src/cli.rs:209]: Nice! A future idea: should we have dry-run be a common flag to all subcommands? Then I guess it is a higher scoped flag in the CLI struct, but it should be settable on the commandline either as `git-toprepo --dry-run <command>` or `git-toprepo <command> --dry-run`.

I think many CLI parsers calls this a global option. @albinvass what do you think? I think we had different perspectives on a similar question a year back or so.

File: src/cli.rs [hunk 2]
```diff
@@ -193,6 +193,78 @@ pub enum Commands {
...
+    #[arg(long, short = 'd')]
```
Comment (stagnation) [inline src/cli.rs:210]: To make sure: is this the common short flag? They are generally discouraged so I would not add them until we are certain we want them.

File: src/cli.rs [hunk 3]
```diff
@@ -193,6 +193,78 @@ pub enum Commands {
...
+
+    /// Fetch objects even if they already exist locally.
+    #[arg(long)]
+    pub refetch: bool,
+
+    /// Exclude paths for this invocation.
+    #[arg(long = "exclude", short = 'X', value_name = "PATHS")]
+    pub exclude: Vec<String>,
+
+    /// Unsupported in git-toprepo's LFS wrapper.
```
Comment (stagnation) [inline src/cli.rs:233]: Thanks, do you want to elaborate on why we eat them? For copy-paste compatibility and the `git toprepo git lfs` techinque? Is a warning raised or is it silently ignored?

Comment (stagnation) [reply src/cli.rs:233]: It is technically supported as you have implemented it, maybe "unused" is better? With the comment as to why and what happens.

File: src/cli.rs [hunk 4]
```diff
@@ -193,6 +193,78 @@ pub enum Commands {
...
+    #[arg(long, short = 'I', hide = true, value_name = "PATHS")]
+    pub include: Vec<String>,
+
+    /// Unsupported in git-toprepo's LFS wrapper.
+    #[arg(long, hide = true)]
+    pub json: bool,
+
+    /// Monorepo path(s) whose LFS objects should be fetched.
+    #[arg(value_name = "PATH", required = true)]
+    pub paths: Vec<PathBuf>,
```
Comment (stagnation) [inline src/cli.rs]: would you move this above the Unsupported flags, to group them you can document that whole class of flags with just one comment.

File: src/cli.rs [hunk 5]
```diff
@@ -193,6 +193,78 @@ pub enum Commands {
...
+}
+
+impl LfsFetch {
+    pub fn validate(&self) -> anyhow::Result<()> {
```
Comment (stagnation) [inline src/cli.rs:252]: Thanks! This answers and encodes many of my questions above :)

File: src/lfs.rs [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+        return false;
+    }
+    let subcommand = subcommand.to_ascii_lowercase();
+    let command_name = Path::new(tokens[0])
+        .file_name()
+        .and_then(|name| name.to_str())
+        .unwrap_or(tokens[0]);
+    let matches_git_toprepo = |tokens: &[&str]| {
+        tokens.len() >= 2
+            && command_name.eq_ignore_ascii_case("git-toprepo")
```
Comment (stagnation) [inline src/lfs.rs:49]: Beautiful! I love it, unfortunately line breaks like these are rare but precisely as I would have them

File: src/lfs.rs [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .trace_command(crate::command_span!("git lfs version"))
+        .safe_status();
+    match status {
+        Ok(status) if status.success() => Ok(()),
+        Ok(status) => {
+            log::info!("Use: `sudo apt install git-lfs`");
+            bail!(
+                "Git LFS is required for `git toprepo lfs fetch`, but `git lfs version` failed.\n\
```
Comment (stagnation) [inline src/lfs.rs]: minor: try to avoid writing the failing command into the message. As then you must update two locations when you change it. This is better split into one print that essentially says "subprocess {args} failed", then the error message: often "git lfs: command not found"
then finally a helpful message that we can give, this is typically done with the error context paths, but be warned that we use them quite inconsistently in here.

File: src/lfs.rs [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+            .get(3)
+            .is_some_and(|cmd| cmd.eq_ignore_ascii_case(&subcommand))
+    {
+        return true;
+    }
+    false
+}
+
+pub fn ensure_git_lfs_available(repo_worktree: &Path) -> Result<()> {
+    let status = Command::new("git")
```
Comment (stagnation) [inline src/lfs.rs]: minor: would it be better to use the hyphenated form? The real `git-lfs` tool?
Rather than relying on the `git` dispatch to `git-lfs`?

File: src/lfs.rs [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+Install Git LFS and ensure `git lfs version` works.\n\
+Status: {status}"
+            )
+        },
+        Err(err) => {
+            log::info!("Use: `sudo apt install git-lfs`");
+            bail!(
+                "Git LFS is required for `git toprepo lfs fetch`, but `git lfs version` failed.\n\
+Install Git LFS and ensure `git lfs version` works.\n\
+Error: {err}"
```
Comment (stagnation) [inline src/lfs.rs]: I have a feeling you can match to this status, then make it an error and handle both errors with a single context statement.

File: src/lfs.rs [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+            cmd.arg("--prune");
+        }
+        if options.recent {
+            cmd.arg("--recent");
+        }
+        if options.refetch {
+            cmd.arg("--refetch");
+        }
+        for exclude in &options.exclude {
+            cmd.arg("-X").arg(exclude);
```
Comment (stagnation) [inline src/lfs.rs]: Is there a long flag we can use instead? make `strace` and such easier to read for the curious yet uninitiated readers.

File: src/lfs.rs [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+        }
+        cmd.arg("-I").arg(target.include_path.to_string());
```
Comment (stagnation) [inline src/lfs.rs]: Is there a long flag here?

File: src/lfs.rs [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+    }
+    Ok(())
+}
+
+fn warn_if_lfs_filter_bypass_toprepo(repo: &Path, key: &str, subcommand: &str) -> Result<()> {
+    for value in crate::git::git_config_get_all(repo, key)? {
+        if !is_lfs_filter_through_toprepo(&value, subcommand) {
+            match subcommand {
+                "smudge" => log::warn!(
+                    "warning: {key} is configured as `{value}`, which bypasses git-toprepo.\n\
```
Comment (stagnation) [inline src/lfs.rs:175]: Thanks, though is this a warning or would you feel safe in creating a hard error?

File: src/lfs.rs [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+In an emulated monorepo this may fetch LFS objects from the wrong remote. Configure the LFS smudge filter to go through git-toprepo when using git-toprepo LFS support."
+                ),
+                "filter-process" => log::warn!(
+                    "warning: {key} is configured as `{value}`, which bypasses git-toprepo.\n\
+Git's long-running filter process can take precedence over filter.lfs.smudge, so this may bypass git-toprepo-aware LFS handling."
```
Comment (stagnation) [inline src/lfs.rs:180]: That is just to say: in large organizations warnings like these are typically not heeded and end up with support tickets. Whereas errors are easier to understand, and makes the support ticket faster.

But of course, the underlying protocol so to speak works either way, it is just that the user almost never wants the bad mode of operation.

File: src/lfs.rs [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+        }
+    }
+    Ok(())
+}
+
+fn default_fetch_url(repo: &gix::Repository) -> Result<gix::Url> {
+    Ok(repo
+        .find_default_remote(gix::remote::Direction::Fetch)
+        .context("Default git-remote not found")?
+        .context("Bad default git-remote")?
```
Comment (stagnation) [inline src/lfs.rs:195]: Do we need two contexts?

File: src/lfs.rs [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+    let path = path.to_string_lossy();
+    if path.contains('*') || path.contains('?') || path.contains('[') || path.contains(']') {
+        bail!(
+            "`git toprepo lfs fetch` currently expects literal paths, not glob patterns: `{path}`"
+        );
+    }
+    Ok(())
+}
+
+fn repo_relative_git_path(worktree: &Path, path: &Path) -> Result<GitPath> {
```
Comment (stagnation) [inline src/lfs.rs:271]: ack. though I actually thought gix had a helper like this. Good that you create one.

File: README.md [hunk 16]
```diff
@@ -31,6 +31,11 @@ use `git toprepo config bootstrap` afterwards to create one.
 
 `git toprepo fetch [<remote>]` fetches from a `remote` and performs the submodule expansion.
 
+`git toprepo lfs fetch [OPTIONS] <PATH>...` routes each monorepo path to the
+correct repository remote and runs `git lfs fetch` with path includes.
+This avoids fetching LFS objects from the wrong remote when working in an
```
Comment (stagnation) [inline README.md]: Thanks an techincally true, but the status quo should be that it works. You should not document that you have solved all bugs, rather express important features. I would skip line 36 entirely, for those who have need seen any bugs it just sows doubt. They are already happy that git-lfs works with git-toprepo (or the other way around)

File: README.md [hunk 17]
```diff
@@ -74,6 +80,35 @@ there is no ancestor that have been expanded into a monocommit.
...
+* `git toprepo lfs fetch -X "*.tmp" path/to/file.bin`
+
+Notes and constraints:
+
+* `git lfs` must be installed and working (`git lfs version`).
+* Paths are treated as literal paths (glob-like path arguments are rejected).
+* Relative paths are accepted, including from subdirectories.
+* Nested submodules use the deepest matching submodule path.
+* If a submodule URL from `.gitmodules` is not configured in `.gittoprepo.toml`,
+  the command fails instead of falling back to a wrong remote.
```
Comment (stagnation) [inline README.md]: Likewise I'm unsure about the "instead of" clause here. Maybe it is relevant maybe not.

File: README.md [hunk 18]
```diff
@@ -74,6 +80,35 @@ there is no ancestor that have been expanded into a monocommit.
...
+* Unsupported flags are rejected: `--all`, `--stdin`, `--include/-I`, `--json`.
```
Comment (stagnation) [inline README.md:102]: Great!

File: README.md [hunk 19]
```diff
@@ -74,6 +80,35 @@ there is no ancestor that have been expanded into a monocommit.
...
+
+Basic usage:
+
+* `git toprepo lfs fetch path/to/file.bin`
+* `git toprepo lfs fetch subrepo/a.bin subrepo/b.bin`
+* `git toprepo lfs fetch --dry-run path/to/file.bin`
+* `git toprepo lfs fetch --prune --recent --refetch path/to/file.bin`
+* `git toprepo lfs fetch -X "*.tmp" path/to/file.bin`
+
+Notes and constraints:
```
Comment (stagnation) [inline README.md:94]: Thanks

**Review by stagnation — COMMENTED** [review]
really nice!

File: src/cli.rs (hunk 2 above)
Comment (detelinamarinova) [reply src/cli.rs:210]: -d is the short flag used by git lfs fetch. I tried to keep the wrapper aligned with the existing Git LFS interface so users familiar with git lfs can use the same commands and flags with git-toprepo lfs where applicable.

File: src/cli.rs (hunk 1 above)
Comment (detelinamarinova) [reply src/cli.rs:209]: Happy to refactor it into a global option if you decide you want that.

File: src/cli.rs (hunk 3 above)
Comment (detelinamarinova) [reply src/cli.rs:233]: The intention was copy-paste compatibility with git lfs fetch: we parse known Git LFS options that do not fit this wrapper, so users get a deliberate error instead of an unknown-argument error.

File: src/lfs.rs (hunk 8 above)
Comment (detelinamarinova) [reply src/lfs.rs]: Yes, that is a good point. I will change it.

File: src/lfs.rs (hunk 10 above)
Comment (detelinamarinova) [reply src/lfs.rs]: Yes, git-lfs supports the long flag too. I will replace this.

File: src/lfs.rs (hunk 11 above)
Comment (detelinamarinova) [reply src/lfs.rs]: Yes, there is. Will replace this too.

File: src/lfs.rs (hunk 12 above)
Comment (detelinamarinova) [reply src/lfs.rs:175]: I would keep it as a warning since users may have standard Git LFS setup configured globally even when they are not working with LFS objects in this particular repo. Making it a hard error could block otherwise valid git-toprepo fetch workflows.

Also, until there is git-toprepo-aware smudge/filter-process wrappers, the safest workaround I know of is to skip automatic smudge during checkout/reset and fetch the needed LFS objects explicitly:
GIT_LFS_SKIP_SMUDGE=1 git reset --hard origin/master
then:
git toprepo lfs fetch <path>
and:
git lfs checkout <path>

File: src/lfs.rs (hunk 13 above)
Comment (detelinamarinova) [reply src/lfs.rs:180]: Would it make sense to keep this as a warning for now and switch it to a hard error once we have a git-toprepo-aware smudge/filter-process wrapper? I can start working on the smudge wrapper next together with the pull wrapper.

File: src/lfs.rs (hunk 14 above)
Comment (detelinamarinova) [reply src/lfs.rs:195]: Yes, since find_default_remote() returns Option<Result<Remote, Error>>, they are handling different layers: None and Err

File: src/lfs.rs (hunk 15 above)
Comment (detelinamarinova) [reply src/lfs.rs:271]: There is gix-path but it is limited

File: src/cli.rs (hunk 1 above)
Comment (moroten) [reply src/cli.rs:209]: What does `git-toprepo --dry-run version` mean? I'm sceptical to add it at global scope. Would `--force` also be a candidate?

File: src/cli.rs (hunk 2 above)
Comment (moroten) [reply src/cli.rs:210]: That is a good plan. Maybe add a comment above.
```suggestion
    // Using `-d` like in git-lfs, not `-n`.
    #[arg(long, short = 'd')]
```

File: src/cli.rs (hunk 3 above)
Comment (moroten) [reply src/cli.rs:233]: That is very good. Would be fun to use a `UnimplementedArg` wrapper class which parses into an error on clap-parse level, but that can be done as a followup.

File: src/cli.rs (hunk 1 above)
Comment (stagnation) [reply src/cli.rs:209]: Good point

File: src/lfs.rs (hunk 13 above)
Comment (stagnation) [reply src/lfs.rs:180]: yes, I just want to raise the question. Almost always the code as written is right, but one the off chance that there is an oversight this is the best process to find those issues. And I also learn more by challenging assumptions.
I appreciate your answers and they are in line with what I expected :)

File: src/lfs.rs (hunk 14 above)
Comment (stagnation) [reply src/lfs.rs:195]: argh. and conveniently they both have the same context trait.
Would you please add a short "NB: " to illuminate this?

**Review by moroten — COMMENTED** [review]
I will continue my review, have not gone through all the code yet...

File: src/cli.rs (hunk 2 above)
Comment (stagnation) [reply src/cli.rs:210]: FTR: That push has a different short flag.
>   -n, --dry-run    Print the push commands to stdout but do not execute them

This is part of the issue with short flags, I don't like them and don't envy you having to juggle it all.

File: src/lfs.rs [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+            .trace_command(crate::command_span!("git-lfs fetch"))
+            .safe_status()?
+            .check_success()
+            .with_context(|| {
+                if options.dry_run {
+                    format!(
+                        "`git-lfs fetch --dry-run` failed for `{}`. \
+Your installed Git LFS version may not support `--dry-run`.",
+                        target.include_path
+                    )
```
Comment (detelinamarinova) [inline src/lfs.rs:160]: Added a check/message for --dry-run failures since it isn't available on older git-lfs versions.
Apt-installed version doesn't support it

File: src/cli.rs (hunk 2 above)
Comment (moroten) [reply src/cli.rs]: Removed the short flag.
